library/core/src/fmt/mod.rs

```diff
@@ -188,8 +188,28 @@ pub trait Write {
     /// assert_eq!(&buf, "world");
     /// ```
     #[stable(feature = "rust1", since = "1.0.0")]
-    fn write_fmt(mut self: &mut Self, args: Arguments<'_>) -> Result {
-        write(&mut self, args)
+    fn write_fmt(&mut self, args: Arguments<'_>) -> Result {
+        // We use a specialization for `Sized` types to avoid an indirection
+        // through `&mut self`
+        trait SpecWriteFmt {
+            fn spec_write_fmt(self, args: Arguments<'_>) -> Result;
+        }
+
+        impl<W: Write + ?Sized> SpecWriteFmt for &mut W {
+            #[inline]
+            default fn spec_write_fmt(mut self, args: Arguments<'_>) -> Result {
+                write(&mut self, args)
+            }
+        }
+
+        impl<W: Write> SpecWriteFmt for &mut W {
+            #[inline]
+            fn spec_write_fmt(self, args: Arguments<'_>) -> Result {
+                write(self, args)
+            }
+        }
+
```

Comment:
This could also have a specialization for `dyn Write` itself:

```suggestion

        impl SpecWriteFmt for &mut (dyn Write + '_) {
            #[inline]
            fn spec_write_fmt(self, args: Arguments<'_>) -> Result {
                write(self, args)
            }
        }

```

... and probably  `dyn Write + Send` and `dyn Write + Send + Sync` as well, because those are considered different types for the sake of specialization.

Reply:
I think that `dyn Write` already has a `Write` impl that goes through dynamic dispatch, so such a specialization would actually never be used ?

Reply:
Yes, that's correct, disregard my comment.